### ISSUE TITLE
labelGrouping "residue" not working

### ISSUE THREAD

**Issue by nvelden** [nglviewer/ngl#837]
If I add a label to the below example it works fine but if I change labelGrouping to "residue" the labels are not displayed. Do I do something wrong?

https://codepen.io/pen?&editors=001

```
stage.loadFile("data://1blu.mmtf").then(function (o) {
  o.addRepresentation("cartoon", { color: "bfactor" })
  o.addRepresentation("label", {
    labelType: "text",
    labelGrouping: 'atom',
    labelText: { 19: 'foo', 21: "ba"},
  })
  o.autoView()
})
```

**Comment by fredludlow** [comment]
I can see why you wrote that, but unfortunately that's not how it works at the moment (though it would make a lot of sense if it did). Currently `labelText` is always keyed on `atomIndex` (the internal, 0-based index, which is the order they're parsed from the file). I think it would be nicer if you could specify selection strings as keys and and labels as values (would welcome a PR in this direction), but the easiest thing now is probably to figure out the atomIndexes you want labelled and construct labelText appropriately. Quick example here (this doesn't think too much about multiple models, alt-locs or anything else sophisticated)

https://codepen.io/fludlow/pen/LYRKgeN?editors=1111



```javascript
var residueLabels = {
    '1:A.CA': 'Foo', 
    '2:A.CA': 'Bar'
}

stage.loadFile("rcsb://1blu.mmtf").then(function (o) {
  o.addRepresentation("line", { color: "bfactor" })
  var sele = new NGL.Selection("")
  var view = o.structure.getView(sele)
  var labelText = { }
  for (seleString in residueLabels) {
      sele.setString(seleString)
      var atomIndex = view.getAtomIndices()[0]
      if (atomIndex !== undefined) {
          labelText[atomIndex] = residueLabels[seleString]
      }
  }
    
  o.addRepresentation("label", {
    labelType: "text",
    labelText: labelText,
  })
  o.autoView('3:A')
})
```

**Comment by fredludlow** [comment]
> How do I show a background for only the "atoms" or "residues" that have a label?

Quickest way is probably to set a selection on the representation, see updated example: https://codepen.io/fludlow/pen/LYRKgeN?editors=1111

**Comment by fredludlow** [comment]
> How do I show a background for only the "atoms" or "residues" that have a label?
BTW, I think this is currently a bug (showing background where label is the empty string)

**Comment by nvelden** [comment]
Thanks a lot for your answers! I would love to have the possibility where you can select a residue and add a custom text. Using the x-, y-, z-Offset you can then exactly position the label where you want. In this way you could achieve the same as the example below:

https://codepen.io/pen?&editors=001

but in (my opinion) an easier way.

**Comment by fredludlow** [comment]
Agreed, I'm unlikely to have any time to implement this (I'm lucky when I get the odd few days to catch up with issues, PRs etc) but I'm *very* happy to review PRs, offer directions etc if someone wants to have a go :)

Note that it's not very pretty but you can just add a separate representation for each label using: `{labelType: 'format', labelFormat: 'Your label here', sele: '1:A.CA'}`

**Comment by nvelden** [comment]
Thanks again! This is a nice solution for now. I would love to work on it but I am not great with JavaScript. I am actually working on a Shiny application and library such that people can use NGL.js directly in R.

**Comment by fredludlow** [comment]
That sounds cool, if you're planning on releasing it let me know and I can add a link to the README

**Comment by fredludlow** [comment]
Just revisiting this - there are also some issues around the type of the `labelText` parameter. The docstring suggests `string[]`, the type definition says `string` and the LabelFactory constructor expects `{ [k: number]: string }`, which is, technically, compatible with `string` I guess?! (you can index a string to get a string...).

**Comment by nvelden** [comment]
> 
> 
> That sounds cool, if you're planning on releasing it let me know and I can add a link to the README

I have finally released the package.  You can find it in the link below: 
-https://nvelden.github.io/NGLVieweR/

I also made a web application (still in progress) using R Shiny that incorporates most of the functions of the package:
-https://www.nvelden.com/shiny/shiny_NGLVieweR/

I would be happy with any feedback / suggestions and honored if you would list it in the README.
